text/0192-remove-constructs-compat.md

```diff
@@ -0,0 +1,1095 @@
+`constructs.Construct` but this is actually not very prevalent in the framework
+code because the majority of constructs actually extend `core.Resource`.
+
+Alternatives considered:
+
+- **Do nothing in 1.x**: will incur an ongoing maintenance cost of the 1.x -> 2.x
+  merge conflicts.
+- **Automatic merge resolution and import organization**: requires research and
+  development, not necessarily worth it.
+
+## 10-CONSTRUCT-NODE
+
+Since we won't be able to release additional major versions of the
+"constructs" library (in order to ensure interoperability between domains is always
+possible), we need to closely examine the API of this library.
+
+In particular, the API of the `Construct` class, which is the base of all
+constructs in all CDKs, should be as small as possible in order not to "pollute"
+domain-specific APIs introduced in various domains.
+
+In many cases (cdk8s, cdktf), constructs are *generated* on-demand from
+domain-specific API specifications. In such cases, we need to ensure that the
+API in `Construct` does not conflict with generated property names or methods.
+
+The current API of `Construct` in the base class (2.x, 3.x) only includes a
+few protected `onXxx` methods (`onPrepare`, `onValidate` and
+`onSynthesize`). Those methods will be removed in 10.x
+([prepare](#06-no-prepare) and [synthesize](#07-no-synthesize) are no longer
+supported and [validate](#08-validation) will be supported through
+`addValidation()`).
+
+In AWS CDK 1.x the construct API is available under `myConstruct.node`. This
+API has been intentionally removed when we extracted "constructs" from the AWS CDK
+in order to allow the compatibility layer in AWS CDK 1.x to use the same property name
+and expose the shim type (jsii does not allow changing the type of a property in a subclass). 
+
+The base library currently offers `Node.of(scope)` as an alternative - but this API is cumbersome
+to use and not discoverable. In evidence, in CDK for Terraform, they chose to offer [`constructNode`]
+in `TerraformElement` as a sugar for `Node.of()`.
+
+[`constructNode`]: https://github.com/hashicorp/terraform-cdk/blob/5becfbc699180adfe920dec794200bbf56dda0a7/packages/cdktf/lib/terraform-element.ts#L21
+
+Another downside of `Node.of()` is that it means that the `IConstruct` interface
+is now an empty interface, which is a very weak type in TypeScript due to
+structural typing (it's structurally identical to `any`).
+
+As we evaluate this use case for constructs 10.x, we would like to restore the
+ability to access the construct API from a property of `Construct`, and use that
+property as the single marker that represents a construct type (`IConstruct`).
+
+To reduce the risk of naming conflicts (e.g. see [Terraform CDK
+issue](https://github.com/hashicorp/terraform-cdk/pull/230)) between `node` and
+domain-specific APIs, we propose to introduce this API under the name
+`construct` (of type `Node`).
```

Comment:
So... let me summarize this: all classes that extend `Construct` will have a `construct` property that returns an instance of `Node`? I would naturally have expected `construct` to return an instance of `Construct` (basically, `this`), and the naming you're proposing strikes me as surprising and counter-intuitive.

Reply:
The property name should be considered in the context of its usage.

Do you find this counterintuitive:

```ts
bucket.construct.addDependency(role);
```

As oppose to:

```ts
bucket.node.addDependency(role);
```

I actually think the former is much more intuitively and reads much better.

We actually have a naming guideline that says that names should _not_ include the type name because it's redundant information in strongly typed environments.

Do you have an alternative suggestion?

Perhaps we can rename `Node` to something else?

Reply:
As discussed... I don't feel *extremely* strong against this. It does feel counter-intuitive to me but I have no idea whether I'm representative of users or not in this respect.

For the record, I think `this.construct.addDependency(role)` reads **super good**, but for some reason the magic weakens when the receiver is not `this`.

Additionally, I cannot come up with a *better* suggestion for a name... so there's that, too.

Reply:
I know it's a little late, but I don't agree with this rename.

I would expect a field called `construct` to point to an object that IS a Construct. I like what they did in TerraCDK, `constructNode` seems like a good compromise.

Reply:
The main reason we have this property is in order to group all the API of `Construct` under a single namespace so it won't "pollute" the APIs of subclasses, which are oftentimes generated classes in CDKs.

The conceptual mental model is that this is "the construct API". The fact that the type is `Node` is actually not in line with this mental model.

If I am looking at this from a usage point of view, I think `construct` makes more sense:

```ts
this.construct.id
this.construct.path
this.construct.addDependency(other)
this.construct.addMetadata
```

This is actually more natural then:

```ts
this.node.id
this.node.path
...
```

Additionally, we actually have a guideline *against* using the type name in symbol names because type information is an intrinsic part of the API (we prefer `handler: lambda.Function` over `handlerFunction: lambda.Function`).

I would rather rename the `Node` class (maybe `ConstructApi`) instead of renaming this property.

Reply:
For the record, after some discussion I think I would have preferred for this member to be called `constructTree`, as this is the member that consumers use to access the features of the "construct tree" (which is a CDK concept that people are, or should be, familiar with which is distinct from the concept of a "construct").

We get things like:

```ts
// All perfectly natural
bucket.constructTree.findChild()
bucket.constructTree.removeChild()
bucket.constructTree.defaultChild
bucket.constructTree.findAll()
bucket.constructTree.children
bucket.constructTree.addWarning()

// Better than what we have now, this name emphasizes the tree traversal involved in these operations
bucket.constructTree.setContext()
bucket.constructTree.applyAspect()
bucket.constructTree.addDependency() 
 
// These now have not such great names, because they apply to the "current node" and don't really have
// strong tree-ish connotations. Don't think any of them are *horrible* but they are certainly not fantastic.
bucket.constructTree.id
bucket.constructTree.path
bucket.constructTree.uniqueId
```

Reply:
I don't think CDK users are required to understand the tree concept. In fact, we have traditionally tried to talk about "scopes" (I.e "the construct is created within the scope of another construct"). From a CS perspective this is indeed a tree, but I don't see how this adds much value to users.

As mentioned above, the primary reason for this property is to group all the construct APIs under an api namespace so they don't pollute or conflict the public api of subclasses, which is the primary programming pattern we use in CDKs.

Reply:
There's a difference between not needing to know that there is a tree structure involved (until you do), and pretending that the tree structure doesn't exist by studiously avoiding calling it a tree.

You might be under the impression that my argument is one about purity of CS concepts. It is not. 

It is about acknowledging the fact that this structure exists, that it underpins the way the CDK works in a way which you can oftentimes ignore, but sometimes is quite relevant and requires people to understand it (see Stages, Stacks, Aspects, LogicalIDs, default children and escape hatches, see `tree.json` and the VSCode plugin, ...) and that there is a term available which accurately describes the semantics, expectations, and the sort of API that people can expect from this structure that they are already intimately familiar with from other programming tasks.

I think being vague and coy and pretending there is no tree when we all full well know that there is, or using the word "construct" to refer to two distinct concepts, is not going to help users form a mental model.

I'm also not saying they won't be able to. Given enough exposure and enough examples, over time people are going to eventually be trained to write the right code. In that sense, ultimately it doesn't even matter what it's called, be it `node`, `construct`, or `abc`, people will learn to cope. 

------

I think I've laid out my reasoning for why I think it's advantageous to call the beast by its name. It's completely fine if you disagree.

Reply:
Cool, thanks for the effort in explaining.

The plan is to introduce `construct` and deprecate `node` over the `1.x` branch in short order. This will give us some "bake time" of a few months until we completely remove `node` in 2.x. This will allow us to get feedback from users in case the mental model is confusing or insufficient.

Reply:
Just stumbled upon this - thanks for making the effort of changing the name 👍